src/modules/linux-micro/kmod/kmod.c

```diff
@@ -0,0 +1,273 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <libkmod.h>
+#include <stdio.h>
+
+#define SOL_LOG_DOMAIN &_log_domain
+#include "sol-log-internal.h"
+SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "linux-micro-kmod");
+
+#include "sol-platform-linux-micro.h"
+#include "sol-util.h"
+#include "sol-vector.h"
+
+static int
+kmod_apply_value(struct kmod_ctx *ctx, char *mod_name)
+{
+    const int probe_flags = KMOD_PROBE_APPLY_BLACKLIST;
+    struct kmod_list *itr, *modlist = NULL;
+    int r = 0;
+
+    SOL_DBG("load: %s", mod_name);
+
+    r = kmod_module_new_from_lookup(ctx, mod_name, &modlist);
+    if (r < 0) {
+        SOL_ERR("Failed to lookup alias '%s': %m", mod_name);
+        return r;
+    }
+
+    if (!modlist) {
+        SOL_ERR("Failed to find module '%s'", mod_name);
+        return -ENOENT;
+    }
+
+    kmod_list_foreach(itr, modlist) {
+        struct kmod_module *mod;
+        int state, err;
+
+        mod = kmod_module_get_module(itr);
+        state = kmod_module_get_initstate(mod);
+
+        switch (state) {
+        case KMOD_MODULE_BUILTIN:
+            SOL_INF("Module '%s' is builtin", kmod_module_get_name(mod));
+            break;
+
+        case KMOD_MODULE_LIVE:
+            SOL_DBG("Module '%s' is already loaded", kmod_module_get_name(mod));
+            break;
+
+        default:
+            err = kmod_module_probe_insert_module(mod, probe_flags,
+                NULL, NULL, NULL, NULL);
+
+            if (err == 0)
+                SOL_INF("Inserted module '%s'", kmod_module_get_name(mod));
+            else if (err == KMOD_PROBE_APPLY_BLACKLIST)
+                SOL_INF("Module '%s' is blacklisted", kmod_module_get_name(mod));
+            else {
+                SOL_ERR("Failed to insert '%s': %m", kmod_module_get_name(mod));
+                r = err;
+            }
+        }
+
+        kmod_module_unref(mod);
+    }
+
+    kmod_module_unref_list(modlist);
+    return r;
+}
+
+static char *
+remove_leading_whitespace(char *input)
+{
+    while (isspace(*input))
+        input++;
+    return input;
+}
+
+static void
+remove_trailing_whitespace(char *input)
+{
+    char *ptr = input + strlen(input) - 1;
+
+    while (ptr > input && isspace(*ptr)) {
+        *ptr = '\0';
+        ptr--;
+    }
+}
```

Comment:
I'd move these two functions to `sol-util.h` and have them to work with `struct sol_str_slice`, see below where I say to change from `fgets()` to `mmap()`.